mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp

```diff
@@ -1939,6 +1939,132 @@ vectorizeInsertSliceOpPrecondition(tensor::InsertSliceOp sliceOp,
   return success();
 }
 
+namespace {
+enum class ConvOperationKind { Conv, Pool };
+} // namespace
+
+static bool isCastOfBlockArgument(Operation *op) {
+  return isa<CastOpInterface>(op) && op->getNumOperands() == 1 &&
+         isa<BlockArgument>(op->getOperand(0));
+}
+
+// Returns the ConvOperationKind of the op using reduceOp of the generic
+// payload. If it is neither a convolution nor a pooling, it returns
+// std::nullopt.
+//
+// If (region has 2 ops (reduction + yield) or 3 ops (extension + reduction
+// + yield) and rhs is not used) then it is the body of a pooling
+// If conv, check for single `mul` predecessor. The `mul` operands must be
+// block arguments or extension of block arguments.
+// Otherwise, check for one or zero `ext` predecessor. The `ext` operands
+// must be block arguments or extension of block arguments.
+static std::optional<ConvOperationKind>
+getConvOperationKind(Operation *reduceOp) {
+  int numBlockArguments =
+      llvm::count_if(reduceOp->getOperands(), llvm::IsaPred<BlockArgument>);
+
+  switch (numBlockArguments) {
+  case 1: {
```

Comment:
Is there a better way to compute the conv operation kind? Matching the number of block arguments and the definitions reaching the op seems a bit fragile? Could we check for interfaces/op kinds? I think matching the concrete ops would be better, even if the list is large.

Reply:
Thanks for the review!

I am not entirely sure if there's a better way than this to tell if this is conv or pool op. If you've know an approach to do it to make this cleaner, please definitely point out!

Just FYI, this diff shows up as a new implementation but in reality this is barely moving from the previous constructor of the Conv1dGenerator. I meant to make the scope of this PR limited to be refactoring and less of an enhancement. Just so I get unblocked from a crash and move on. Based on this I am prone to treat this one as out of scope area for this PR.